File: src/analysis_py/analysis.py
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

train_df = pd.read_csv("data/titanic/train.csv")

""" CLEANING """

# age, cabin, and embarked have empty entries
print("------ Cleaning the Training Data ------")
print("Pre-cleaned columns in training data")
print(train_df.columns.to_list())
# for age, we will fill it in with the median
print("Filling in nulls in 'Age' and 'Embarked.")
train_df["Age"] = train_df["Age"].fillna(train_df["Age"].median())

# fill in Embarked with the mode
train_df["Embarked"] = train_df["Embarked"].fillna(train_df["Embarked"].mode()[0])

# drop the name, ticker, and cabin column since that won't have any relation to predicting survivability
train_df = train_df.drop(columns=["Name", "Cabin", "Ticket"])
print("Dropped columns 'Name', 'Cabin' and 'Ticket':")
print(train_df.columns.to_list())
# convert categorical variables like Survived, Pclass, Sex, and Embarked as Dummy Variables
train_df = pd.get_dummies(train_df, columns=["Survived", "Pclass", "Sex", "Embarked"])
print("Converted categorical columns in training set to factors.")
print(train_df.columns.to_list())

# checking nulls and shape
print(f"Number of empty entries in the dataframe: {train_df.isnull().any().sum()}")
print("The shape of the training data: ", train_df.shape)

""" EXPLORATION """
print("------ Exploring the Training Data ------")
print("We have checked the numeric values for correlations previously.")
# check the correlation matrix
# print(train_df.corr(numeric_only=True))
# we do not notice any high correlations between the variables

""" MODEL """
print("------ Developing a Logistic Model ------")

print("Extracted the features.")
features = ["Age", "SibSp", "Parch", "Fare", 'Pclass_1', 'Pclass_2', 'Pclass_3',
                    'Sex_female', 'Sex_male', 'Embarked_C', 'Embarked_Q', 'Embarked_S']
model = LogisticRegression()
print("Scaling the predictors for the training model.")
scaler = StandardScaler()
X_scaled = pd.DataFrame(scaler.fit_transform(train_df[features]), columns=features)
print("First three rows of scaled predictors: \n", X_scaled.head(3))
print("Fitting the model: ")
model.fit(X_scaled, train_df['Survived_1'])
print("Intercept: ", model.intercept_)
print("Coefficients: ", model.coef_)

# predict on the train test
Y_pred_train = model.predict(train_df[features])
score = model.score(X_scaled, train_df["Survived_1"])
print("Tested our model on our training set.")
print(f"Score recevied after testing: {score}")

# --- predict on the test set --- #

print("------ Predicting on Test Data ------")
test_df = pd.read_csv("data/titanic/test.csv")

print("The columns in the testing data prior to cleaning for prediction: ", test_df.columns.to_list())

# Fill missing values instead of dropping
test_df["Age"] = test_df["Age"].fillna(test_df["Age"].median())
test_df["Fare"] = test_df["Fare"].fillna(test_df["Fare"].median())
test_df["Embarked"] = test_df["Embarked"].fillna(test_df["Embarked"].mode()[0])
test_df = test_df.drop(columns=["Name", "Cabin", "Ticket"])

print(f"Number of empty entries in the dataframe: {test_df.isnull().any().sum()}")
print("The shape of the testing data: ", test_df.shape)
print("The columns in the testing data prior to cleaning for prediction: ", test_df.columns.to_list())

# converting test_df categorical cols to factors and verifying
test_df = pd.get_dummies(test_df, columns=["Pclass", "Sex", "Embarked"])
print("Converting test category columns to factors.")

# capturing predictions
print("Predicting on the test set... ")
Y_pred_test = model.predict(test_df[features])
print("Predicted!")

# scaling the inputs for the test
print("Scaling the predictors in the test data.")
scaler_test = StandardScaler()
X_scaled_test = pd.DataFrame(scaler.fit_transform(test_df[features]), columns=features)
print("First three rows of scaled predictors: \n", X_scaled_test.head(3))

# checking our model
Y_test_true = pd.read_csv("data/titanic/gender_submission.csv") 
Y_test_true = pd.get_dummies(Y_test_true, columns=["Survived"])
score = model.score(X_scaled_test, Y_test_true["Survived_1"])
print("Testing our model on our testing set.")
print(f"Score recevied after testing on training set: {score}")

File: src/analysis_R/analysis.R
# Load libraries
library(dplyr)
library(caret)
library(broom)

cat("------ Loading Training Data ------\n")
train_df <- read.csv("data/titanic/train.csv")

# ---------- CLEANING ----------
cat("------ Cleaning the Training Data ------\n")
cat("Pre-cleaned columns in training data:\n")
print(colnames(train_df))

# Fill missing Age and Embarked values
cat("Filling in nulls in 'Age' and 'Embarked'.\n")
train_df$Age[is.na(train_df$Age)] <- median(train_df$Age, na.rm = TRUE)
train_df$Embarked[is.na(train_df$Embarked)] <- names(sort(table(train_df$Embarked), decreasing = TRUE))[1]

# Drop unused columns
cat("Dropping columns 'Name', 'Cabin', and 'Ticket'.\n")
train_df <- train_df %>% select(-Name, -Cabin, -Ticket)
print(colnames(train_df))

# Convert categorical variables
cat("Converting categorical columns to factors.\n")
train_df <- train_df %>%
  mutate(
    Survived = factor(Survived),
    Pclass = factor(Pclass),
    Sex = factor(Sex),
    Embarked = factor(Embarked)
  )

features <- c("Age", "SibSp", "Parch", "Fare", "Pclass", "Sex", "Embarked")

cat("Number of empty entries: ", sum(is.na(train_df)), "\n")
cat("Shape of the training data: ", paste(dim(train_df), collapse = " x "), "\n")

# ---------- MODEL ----------
cat("------ Developing a Logistic Model ------\n")

features <- c("Age", "SibSp", "Parch", "Fare", "Pclass", "Sex", "Embarked")

# Build formula dynamically
formula <- as.formula(paste("Survived ~", paste(features, collapse = " + ")))

# Train logistic regression model
cat("Fitting logistic regression model...\n")
model <- suppressWarnings(
  train(formula, data = train_df, method = "glm", family = "binomial")
)

cat("Model training complete.\n")

# Extract and print coefficients + intercept
cat("Intercept and Coefficients:\n")
coefs <- tidy(model$finalModel)
print(coefs)

# Predict on training set
cat("Testing our model on training set...\n")
pred_train <- predict(model, train_df)
acc_train <- mean(pred_train == train_df$Survived)
cat("Accuracy on training data: ", acc_train, "\n")

# ---------- TEST DATA ----------
cat("------ Predicting on Test Data ------\n")
test_df <- read.csv("data/titanic/test.csv")

cat("Columns in testing data prior to cleaning:\n")
print(colnames(test_df))

# Fill missing data
test_df$Age[is.na(test_df$Age)] <- median(test_df$Age, na.rm = TRUE)
test_df$Fare[is.na(test_df$Fare)] <- median(test_df$Fare, na.rm = TRUE)
test_df$Embarked[is.na(test_df$Embarked)] <- names(sort(table(test_df$Embarked), decreasing = TRUE))[1]

# Drop unused columns
test_df <- test_df %>% select(-Name, -Cabin, -Ticket)
cat("Dropped unused columns.\n")

cat("Columns in testing data following cleaning:\n")
print(colnames(test_df))
cat("Number of empty entries: ", sum(is.na(test_df)), "\n")
cat("Shape of the testing data: ", paste(dim(test_df), collapse = " x "), "\n")

# Convert to factors matching training
test_df <- test_df %>%
  mutate(
    Pclass = factor(Pclass, levels = levels(train_df$Pclass)),
    Sex = factor(Sex, levels = levels(train_df$Sex)),
    Embarked = factor(Embarked, levels = levels(train_df$Embarked))
  )

# Predict
cat("Predicting on test data...\n")
pred_test <- predict(model, test_df)
cat("Predicted!\n")

# Evaluate accuracy against gender_submission.csv
Y_test_true <- read.csv("data/titanic/gender_submission.csv")
Y_test_true$Survived <- factor(Y_test_true$Survived, levels = levels(train_df$Survived))

acc_test <- mean(pred_test == Y_test_true$Survived)
cat("Accuracy on predicting test data: ", acc_test, "\n")
